{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c1956dcb",
   "metadata": {},
   "source": [
    "# Creating genes mapping table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5db1e918",
   "metadata": {},
   "source": [
    "# I / Install missing modules & load modules "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "18e11ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra\n",
    "import requests\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87794a8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.error import HTTPError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dabbe5d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "#If this doesn't run install pandarallel in your conda enviroment using conda or pip\n",
    "#https://github.com/nalepae/pandarallel \n",
    "\n",
    "# => Problem to install pandarallel => seems to work with Python version 3.{5, 6, 7, 8}, not 3.9\n",
    "# => Create a new environment containing the same packages, but changing the version of Python : Python 3.8\n",
    "# Then, try to install pandarallel\n",
    "\n",
    "# => For this gene mapping, use the environment named env_aracore_update_gene_mapping => same problem\n",
    "# same problem for all the versions of python...\n",
    "\n",
    "from pandarallel import pandarallel \n",
    "\n",
    "pandarallel.initialize(progress_bar=True)\n",
    "\n",
    "from pandarallel.utils import progress_bars\n",
    "progress_bars.is_notebook_lab = lambda : True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6c8eaff",
   "metadata": {},
   "source": [
    "# II / Load AraCore model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2bb25be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get file from github\n",
    "fileName = 'https://raw.githubusercontent.com/ma-blaetke/CBM_C3_C4_Metabolism/master/data/2018-23-05-mb-genC3.sbml'\n",
    "r = requests.get(fileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "26fa2d6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create model\n",
    "model = cobra.io.read_sbml_model(r.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "929cc8d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>c3_model</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f427cd6d550</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>413</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>572</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Ex_Suc - 1.0*Ex_Suc_reverse_fb96e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Chloroplast, Lumen, Cytosol, Mitochondrion, IntermembraneSpace, Peroxisome</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model c3_model at 0x7f427cd6d550>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29b27c4c",
   "metadata": {},
   "source": [
    "# III / Create Gene Table for AraCore Model  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4256d148",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aracore_ids</th>\n",
       "      <th>aracore_name</th>\n",
       "      <th>aracore_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2G30570</td>\n",
       "      <td>AT2G30570</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATCG00560</td>\n",
       "      <td>ATCG00560</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT4G21280</td>\n",
       "      <td>AT4G21280</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATCG00300</td>\n",
       "      <td>ATCG00300</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AT1G06680</td>\n",
       "      <td>AT1G06680</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ATCG00080</td>\n",
       "      <td>ATCG00080</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AT2G06520</td>\n",
       "      <td>AT2G06520</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AT1G79040</td>\n",
       "      <td>AT1G79040</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AT5G66570</td>\n",
       "      <td>AT5G66570</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AT3G21055</td>\n",
       "      <td>AT3G21055</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ATCG00020</td>\n",
       "      <td>ATCG00020</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ATCG00270</td>\n",
       "      <td>ATCG00270</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ATCG00690</td>\n",
       "      <td>ATCG00690</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ATCG00710</td>\n",
       "      <td>ATCG00710</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATCG00550</td>\n",
       "      <td>ATCG00550</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ATCG00570</td>\n",
       "      <td>ATCG00570</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AT1G67740</td>\n",
       "      <td>AT1G67740</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ATCG00580</td>\n",
       "      <td>ATCG00580</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AT3G50820</td>\n",
       "      <td>AT3G50820</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ATCG00680</td>\n",
       "      <td>ATCG00680</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATCG00280</td>\n",
       "      <td>ATCG00280</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AT1G44575</td>\n",
       "      <td>AT1G44575</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATCG00220</td>\n",
       "      <td>ATCG00220</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ATCG00700</td>\n",
       "      <td>ATCG00700</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   aracore_ids aracore_name aracore_annotations\n",
       "0    AT4G05180    AT4G05180                  {}\n",
       "1    AT2G30570    AT2G30570                  {}\n",
       "2    ATCG00560    ATCG00560                  {}\n",
       "3    AT4G21280    AT4G21280                  {}\n",
       "4    ATCG00300    ATCG00300                  {}\n",
       "5    AT1G06680    AT1G06680                  {}\n",
       "6    ATCG00080    ATCG00080                  {}\n",
       "7    AT2G06520    AT2G06520                  {}\n",
       "8    AT1G79040    AT1G79040                  {}\n",
       "9    AT5G66570    AT5G66570                  {}\n",
       "10   AT3G21055    AT3G21055                  {}\n",
       "11   ATCG00020    ATCG00020                  {}\n",
       "12   ATCG00270    ATCG00270                  {}\n",
       "13   ATCG00690    ATCG00690                  {}\n",
       "14   ATCG00710    ATCG00710                  {}\n",
       "15   ATCG00550    ATCG00550                  {}\n",
       "16   ATCG00570    ATCG00570                  {}\n",
       "17   AT1G67740    AT1G67740                  {}\n",
       "18   ATCG00580    ATCG00580                  {}\n",
       "19   AT3G50820    AT3G50820                  {}\n",
       "20   ATCG00680    ATCG00680                  {}\n",
       "21   ATCG00280    ATCG00280                  {}\n",
       "22   AT1G44575    AT1G44575                  {}\n",
       "23   ATCG00220    ATCG00220                  {}\n",
       "24   ATCG00700    ATCG00700                  {}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create mapping table\n",
    "df_genes_aracore = pd.DataFrame(\n",
    "    {\n",
    "        \"aracore_ids\" : [gene_id.id for gene_id in model.genes],\n",
    "        \"aracore_name\" : [gene_id.name for gene_id in model.genes],\n",
    "        \"aracore_annotations\" : [gene_id.annotation for gene_id in model.genes]\n",
    "    })\n",
    "\n",
    "df_genes_aracore.head(25) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "66be8a3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "804"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_aracore['aracore_ids'].shape[0] #804 genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9269c31b",
   "metadata": {},
   "source": [
    "# IV / Retrieve additional information on genes using the ENSEMBL API "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78ffcfe1",
   "metadata": {},
   "source": [
    "**ENSEMBL** is a database that provides information about **genomic features**. It contains the IDs and the symbol of the genes, as well as the species.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0094a7a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'strand': -1,\n",
       " 'start': 2671523,\n",
       " 'species': 'arabidopsis_thaliana',\n",
       " 'description': 'Oxygen-evolving enhancer protein 3-2, chloroplastic [Source:UniProtKB/Swiss-Prot;Acc:Q41932]',\n",
       " 'object_type': 'Gene',\n",
       " 'display_name': 'PSBQ2',\n",
       " 'Transcript': [{'assembly_name': 'TAIR10',\n",
       "   'biotype': 'protein_coding',\n",
       "   'end': 2673243,\n",
       "   'source': 'araport11',\n",
       "   'is_canonical': 1,\n",
       "   'db_type': 'core',\n",
       "   'Parent': 'AT4G05180',\n",
       "   'Translation': {'end': 2673170,\n",
       "    'id': 'AT4G05180.1',\n",
       "    'length': 230,\n",
       "    'db_type': 'core',\n",
       "    'Parent': 'AT4G05180.1',\n",
       "    'object_type': 'Translation',\n",
       "    'start': 2672093,\n",
       "    'species': 'arabidopsis_thaliana'},\n",
       "   'species': 'arabidopsis_thaliana',\n",
       "   'start': 2671523,\n",
       "   'strand': -1,\n",
       "   'seq_region_name': '4',\n",
       "   'Exon': [{'end': 2673243,\n",
       "     'id': 'AT4G05180.1.exon1',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'seq_region_name': '4',\n",
       "     'db_type': 'core',\n",
       "     'strand': -1,\n",
       "     'start': 2672873,\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'object_type': 'Exon'},\n",
       "    {'assembly_name': 'TAIR10',\n",
       "     'seq_region_name': '4',\n",
       "     'end': 2672635,\n",
       "     'id': 'AT4G05180.1.exon2',\n",
       "     'db_type': 'core',\n",
       "     'object_type': 'Exon',\n",
       "     'start': 2672340,\n",
       "     'strand': -1,\n",
       "     'species': 'arabidopsis_thaliana'},\n",
       "    {'object_type': 'Exon',\n",
       "     'strand': -1,\n",
       "     'start': 2671900,\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'db_type': 'core',\n",
       "     'end': 2672191,\n",
       "     'id': 'AT4G05180.1.exon3',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'seq_region_name': '4'},\n",
       "    {'start': 2671523,\n",
       "     'strand': -1,\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'object_type': 'Exon',\n",
       "     'db_type': 'core',\n",
       "     'end': 2671718,\n",
       "     'id': 'AT4G05180.1.exon4',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'seq_region_name': '4'}],\n",
       "   'id': 'AT4G05180.1',\n",
       "   'logic_name': 'araport11',\n",
       "   'object_type': 'Transcript'},\n",
       "  {'Exon': [{'species': 'arabidopsis_thaliana',\n",
       "     'strand': -1,\n",
       "     'start': 2672873,\n",
       "     'object_type': 'Exon',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'seq_region_name': '4',\n",
       "     'id': 'AT4G05180.1.exon1',\n",
       "     'end': 2673243,\n",
       "     'db_type': 'core'},\n",
       "    {'end': 2672635,\n",
       "     'id': 'AT4G05180.1.exon2',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'seq_region_name': '4',\n",
       "     'db_type': 'core',\n",
       "     'start': 2672340,\n",
       "     'strand': -1,\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'object_type': 'Exon'},\n",
       "    {'seq_region_name': '4',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'id': 'AT4G05180.2.exon3',\n",
       "     'end': 2672191,\n",
       "     'db_type': 'core',\n",
       "     'object_type': 'Exon',\n",
       "     'strand': -1,\n",
       "     'start': 2671784,\n",
       "     'species': 'arabidopsis_thaliana'}],\n",
       "   'id': 'AT4G05180.2',\n",
       "   'seq_region_name': '4',\n",
       "   'logic_name': 'araport11',\n",
       "   'object_type': 'Transcript',\n",
       "   'is_canonical': 0,\n",
       "   'source': 'araport11',\n",
       "   'end': 2673243,\n",
       "   'assembly_name': 'TAIR10',\n",
       "   'biotype': 'protein_coding',\n",
       "   'Parent': 'AT4G05180',\n",
       "   'db_type': 'core',\n",
       "   'start': 2671784,\n",
       "   'species': 'arabidopsis_thaliana',\n",
       "   'strand': -1,\n",
       "   'Translation': {'length': 230,\n",
       "    'id': 'AT4G05180.2',\n",
       "    'end': 2673170,\n",
       "    'Parent': 'AT4G05180.2',\n",
       "    'db_type': 'core',\n",
       "    'start': 2672093,\n",
       "    'species': 'arabidopsis_thaliana',\n",
       "    'object_type': 'Translation'}}],\n",
       " 'assembly_name': 'TAIR10',\n",
       " 'biotype': 'protein_coding',\n",
       " 'seq_region_name': '4',\n",
       " 'id': 'AT4G05180',\n",
       " 'source': 'araport11',\n",
       " 'end': 2673243,\n",
       " 'logic_name': 'araport11',\n",
       " 'db_type': 'core'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#EXAMPLE REQUEST:\n",
    "\n",
    "#Retrieve Information on gene symbol, location (chromosome, start and end), uniprot id/descriptive name using the ENSEMBL REST API\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/lookup\n",
    "\n",
    "gene_id = 'AT4G05180'\n",
    "\n",
    "#ENSEMBL Server Url\n",
    "server = \"https://rest.ensembl.org\"\n",
    "\n",
    "#URL specification to get meta-information of a gene by its id\n",
    "ext = f\"/lookup/id/{gene_id}?expand=1\"\n",
    "  \n",
    "#Request information from URL in json format\n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "\n",
    "#Check if request wad successful\n",
    "if r.ok:\n",
    "    \n",
    "    #Get json (dictionary-like structure) from request\n",
    "    json_anno = r.json()\n",
    "\n",
    "else:\n",
    "\n",
    "    json_anno = None\n",
    "\n",
    "json_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "71d879a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retrieve Information on gene symbol, location (chromosome, start and end), uniprot id/descriptive name using the ENSEMBL REST API\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/lookup\n",
    "\n",
    "def lookup_ensembl_id(gene_id):\n",
    "\n",
    "    #pandas series to hold information\n",
    "    ds_anno = pd.Series(index=['symbol','chr','start', 'end', 'description'])\n",
    "  \n",
    "    #ENSEMBL Server Url\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "  \n",
    "    #URL specification to get meta-information of a gene by its id\n",
    "    ext = f\"/lookup/id/{gene_id}?expand=1\"\n",
    "  \n",
    "  #Request information from URL in json format\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "  \n",
    "  #Check if request wad successful\n",
    "    if r.ok:\n",
    "    \n",
    "        #Get json (dictionary-like structure) from request\n",
    "        json_anno = r.json()\n",
    "\n",
    "        #Save relevant information in pandas series\n",
    "        ds_anno['symbol'] = json_anno.get('display_name', None)\n",
    "        ds_anno['chr'] = json_anno.get('seq_region_name', None)\n",
    "        ds_anno['start'] = json_anno.get('start', None)\n",
    "        ds_anno['end'] = json_anno.get('end', None)\n",
    "        ds_anno['description'] = json_anno.get('description', None)\n",
    "    else:\n",
    "        #Save set pandas series to None if request was not successfull\n",
    "        ds_anno['symbol'] = None\n",
    "        ds_anno['chr'] = None\n",
    "        ds_anno['start'] = None\n",
    "        ds_anno['end'] = None\n",
    "        ds_anno['description'] = None\n",
    "\n",
    "    #return pandas series\n",
    "    return ds_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8c302c3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db_display_name</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>version</th>\n",
       "      <th>info_text</th>\n",
       "      <th>primary_id</th>\n",
       "      <th>display_id</th>\n",
       "      <th>info_type</th>\n",
       "      <th>description</th>\n",
       "      <th>xref_end</th>\n",
       "      <th>xref_identity</th>\n",
       "      <th>ensembl_end</th>\n",
       "      <th>ensembl_identity</th>\n",
       "      <th>score</th>\n",
       "      <th>xref_start</th>\n",
       "      <th>evalue</th>\n",
       "      <th>cigar_line</th>\n",
       "      <th>ensembl_start</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NASC_GENE_ID</th>\n",
       "      <td>NASC Gene ID</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>AT4G05180-TAIR-G</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>photosystem II subunit Q-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ArrayExpress</th>\n",
       "      <td>Expression Atlas</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAIR_LOCUS</th>\n",
       "      <td>TAIR</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>photosystem II subunit Q-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAIR_SYMBOL</th>\n",
       "      <td>TAIR Gene Name</td>\n",
       "      <td>[PSBQ-1, PSBQ-2, PSBQA, PSII-Q]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>PSBQ</td>\n",
       "      <td>PSBQ</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniGene</th>\n",
       "      <td>UniGene</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>At.21853</td>\n",
       "      <td>At.21853</td>\n",
       "      <td>SEQUENCE_MATCH</td>\n",
       "      <td>Oxygen-evolving enhancer protein 3-2</td>\n",
       "      <td>1037.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1037.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>5185.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1037M</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EntrezGene</th>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>825866</td>\n",
       "      <td>825866</td>\n",
       "      <td>DEPENDENT</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KNETMINER_ARA</th>\n",
       "      <td>KNETMINER_ARA</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>DEPENDENT</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               db_display_name  \\\n",
       "dbname                                           \n",
       "NASC_GENE_ID                      NASC Gene ID   \n",
       "ArrayExpress                  Expression Atlas   \n",
       "TAIR_LOCUS                                TAIR   \n",
       "TAIR_SYMBOL                     TAIR Gene Name   \n",
       "UniGene                                UniGene   \n",
       "EntrezGene     NCBI gene (formerly Entrezgene)   \n",
       "KNETMINER_ARA                    KNETMINER_ARA   \n",
       "\n",
       "                                      synonyms version info_text primary_id  \\\n",
       "dbname                                                                        \n",
       "NASC_GENE_ID                                []       0            AT4G05180   \n",
       "ArrayExpress                                []       0            AT4G05180   \n",
       "TAIR_LOCUS                                  []       0            AT4G05180   \n",
       "TAIR_SYMBOL    [PSBQ-1, PSBQ-2, PSBQA, PSII-Q]       0                 PSBQ   \n",
       "UniGene                                     []       0             At.21853   \n",
       "EntrezGene                                  []       0               825866   \n",
       "KNETMINER_ARA                               []       0            AT4G05180   \n",
       "\n",
       "                     display_id       info_type  \\\n",
       "dbname                                            \n",
       "NASC_GENE_ID   AT4G05180-TAIR-G          DIRECT   \n",
       "ArrayExpress          AT4G05180          DIRECT   \n",
       "TAIR_LOCUS            AT4G05180          DIRECT   \n",
       "TAIR_SYMBOL                PSBQ          DIRECT   \n",
       "UniGene                At.21853  SEQUENCE_MATCH   \n",
       "EntrezGene               825866       DEPENDENT   \n",
       "KNETMINER_ARA         AT4G05180       DEPENDENT   \n",
       "\n",
       "                                        description  xref_end  xref_identity  \\\n",
       "dbname                                                                         \n",
       "NASC_GENE_ID             photosystem II subunit Q-2       NaN            NaN   \n",
       "ArrayExpress                                              NaN            NaN   \n",
       "TAIR_LOCUS               photosystem II subunit Q-2       NaN            NaN   \n",
       "TAIR_SYMBOL                                               NaN            NaN   \n",
       "UniGene        Oxygen-evolving enhancer protein 3-2    1037.0          100.0   \n",
       "EntrezGene                                     None       NaN            NaN   \n",
       "KNETMINER_ARA                                  None       NaN            NaN   \n",
       "\n",
       "               ensembl_end  ensembl_identity   score  xref_start  evalue  \\\n",
       "dbname                                                                     \n",
       "NASC_GENE_ID           NaN               NaN     NaN         NaN     NaN   \n",
       "ArrayExpress           NaN               NaN     NaN         NaN     NaN   \n",
       "TAIR_LOCUS             NaN               NaN     NaN         NaN     NaN   \n",
       "TAIR_SYMBOL            NaN               NaN     NaN         NaN     NaN   \n",
       "UniGene             1037.0              96.0  5185.0         1.0     NaN   \n",
       "EntrezGene             NaN               NaN     NaN         NaN     NaN   \n",
       "KNETMINER_ARA          NaN               NaN     NaN         NaN     NaN   \n",
       "\n",
       "              cigar_line  ensembl_start  \n",
       "dbname                                   \n",
       "NASC_GENE_ID         NaN            NaN  \n",
       "ArrayExpress         NaN            NaN  \n",
       "TAIR_LOCUS           NaN            NaN  \n",
       "TAIR_SYMBOL          NaN            NaN  \n",
       "UniGene            1037M            1.0  \n",
       "EntrezGene           NaN            NaN  \n",
       "KNETMINER_ARA        NaN            NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#EXAMPLE REQUEST\n",
    "\n",
    "#Retrieve Information on cross references for each gene symbol (NCBI gene id)\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/xref_id\n",
    "\n",
    "gene_id = 'AT4G05180'\n",
    "\n",
    "#ENSEMBL Server Url\n",
    "server = \"https://rest.ensembl.org\"\n",
    "\n",
    "#URL specification to get meta-information of a gene by its id\n",
    "ext = f\"/xrefs/id/{gene_id}?\"\n",
    "  \n",
    "#Request information from URL in json format\n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "\n",
    "#Check if request was successful\n",
    "if r.ok:\n",
    "    \n",
    "    #Get json (dictionary-like structure) from request\n",
    "    json_anno = r.json() # json format file for annotations containing r variable content\n",
    "\n",
    "    #Convert json to pandas dataframe format to easily retrieve information\n",
    "    df_anno = pd.DataFrame(json_anno) # dataframe containing annotations\n",
    "    df_anno.set_index('dbname', inplace=True) # delete the index on the left of the dataframe\n",
    "else:\n",
    "     df_anno = None\n",
    "\n",
    "df_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c4d3ef32",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retrieve Information on cross references for each gene symbol (NCBI gene id)\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/xref_id\n",
    "\n",
    "def xrefs_ensembl_id(gene_id):\n",
    "  \n",
    "  #pandas series to hold information\n",
    "    ds_anno = pd.Series(index=['ncbigene'])\n",
    "  \n",
    "  #ENSEMBL Server Url\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "  #URL specification to get meta-information of a gene by its id\n",
    "    ext = f\"/xrefs/id/{gene_id}?\"\n",
    "  \n",
    "  #Request information from URL in json format\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "  \n",
    "  #Check if request wad successful\n",
    "    if r.ok:\n",
    "    \n",
    "    #Get json (dictionary-like structure) from request\n",
    "        json_anno = r.json()\n",
    "\n",
    "        #Convert json to pandas dataframe format\n",
    "        df_anno = pd.DataFrame(json_anno)\n",
    "        df_anno.set_index('dbname', inplace=True)\n",
    "\n",
    "        #Save relevant information in pandas series\n",
    "        if 'EntrezGene' in df_anno.index:\n",
    "              if 'primary_id' in df_anno.columns:\n",
    "                    ds_anno['ncbigene'] = str(df_anno.loc['EntrezGene','primary_id'])\n",
    "\n",
    "    else:\n",
    "        ds_anno['ncbigene'] = None\n",
    " \n",
    "    return ds_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8d07497b",
   "metadata": {},
   "outputs": [
    {
     "ename": "SystemError",
     "evalue": "Python version should be 3.{5, 6, 7}",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mSystemError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-a3f59e75c06b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#Request information for each gene using xrefs_ensembl_id() function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf_genes_xref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'aracore_ids'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparallel_apply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mxrefs_ensembl_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/env_aracore_update_gene_mapping/lib/python3.8/site-packages/pandarallel/pandarallel.py\u001b[0m in \u001b[0;36mclosure\u001b[0;34m(data, func, *args, **kwargs)\u001b[0m\n\u001b[1;32m    424\u001b[0m         \u001b[0mqueue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmanager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mQueue\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    425\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 426\u001b[0;31m         workers_args, chunk_lengths, input_files, output_files = get_workers_args(\n\u001b[0m\u001b[1;32m    427\u001b[0m             \u001b[0muse_memory_fs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    428\u001b[0m             \u001b[0mnb_requested_workers\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/env_aracore_update_gene_mapping/lib/python3.8/site-packages/pandarallel/pandarallel.py\u001b[0m in \u001b[0;36mget_workers_args\u001b[0;34m(use_memory_fs, nb_workers, progress_bar, chunks, worker_meta_args, queue, func, args, kwargs)\u001b[0m\n\u001b[1;32m    279\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mOSError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    280\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 281\u001b[0;31m         workers_args = [\n\u001b[0m\u001b[1;32m    282\u001b[0m             (\n\u001b[1;32m    283\u001b[0m                 \u001b[0minput_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/env_aracore_update_gene_mapping/lib/python3.8/site-packages/pandarallel/pandarallel.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    288\u001b[0m                 \u001b[0mprogress_bar\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mPROGRESS_IN_WORKER\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m                 dill.dumps(\n\u001b[0;32m--> 290\u001b[0;31m                     progress_wrapper(\n\u001b[0m\u001b[1;32m    291\u001b[0m                         \u001b[0mprogress_bar\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mPROGRESS_IN_FUNC\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mqueue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchunk_length\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m                     )(func)\n",
      "\u001b[0;32m~/miniconda3/envs/env_aracore_update_gene_mapping/lib/python3.8/site-packages/pandarallel/pandarallel.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(func)\u001b[0m\n\u001b[1;32m    198\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    199\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mprogress_bar\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 200\u001b[0;31m             wrapped_func = inline(\n\u001b[0m\u001b[1;32m    201\u001b[0m                 \u001b[0mprogress_pre_func\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    202\u001b[0m                 \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/env_aracore_update_gene_mapping/lib/python3.8/site-packages/pandarallel/utils/inliner.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     30\u001b[0m         \u001b[0mpython_version\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mversion_info\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mpython_version\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmajor\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m3\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mpython_version\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mminor\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m6\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m7\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 32\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mSystemError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Python version should be 3.{5, 6, 7}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mSystemError\u001b[0m: Python version should be 3.{5, 6, 7}"
     ]
    }
   ],
   "source": [
    "#Request information for each gene using xrefs_ensembl_id() function\n",
    "df_genes_xref = df_genes_aracore['aracore_ids'].parallel_apply(lambda x: xrefs_ensembl_id(x)) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bba149a1",
   "metadata": {},
   "source": [
    "**Concerning the cell below**:\n",
    "\n",
    "We write a REQUEST, which contains the URL of the ENSEMBL gene ID, and the Output format, under the following form :\n",
    "REQUEST(URL, Output format)\n",
    "\n",
    "This URL is constituted into 2 parts:\n",
    "- the base of the URL, which refers to the database ENSEMBL\n",
    "- the keywords, which will be translated into MySQL\n",
    "\n",
    "These 2 parts of the REQUEST are together converted into the Output format. \n",
    "\n",
    "This Output format is JSON, similar to a dictionary in Python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9f09316",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Request information for each gene using lookup_ensembl_id() function\n",
    "df_genes_lookup = df_genes_aracore['aracore_ids'].parallel_apply(lambda x: lookup_ensembl_id(x)) #This may take a few minutes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b6ddd88",
   "metadata": {},
   "source": [
    "# V / Merge information into a single mapping table "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "89432cee",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_genes_lookup' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-60aae18e7ced>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#Concat dataframe df_genes_aracore, df_genes_lookup, df_genes_xref\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf_genes_aracore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_genes_lookup\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_genes_xref\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_genes_lookup' is not defined"
     ]
    }
   ],
   "source": [
    "#Concat dataframe df_genes_aracore, df_genes_lookup, df_genes_xref\n",
    "df_genes_aracore = pd.concat([df_genes_aracore, df_genes_lookup, df_genes_xref], axis=1) \n",
    "df_genes_aracore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "abe2a62e",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'description'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'description'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-0df3ce687e44>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#e.g. \"Oxygen-evolving enhancer protein 3-2, chloroplastic [Source:UniProtKB/Swiss-Prot;Acc:Q41932]\" -> \"Oxygen-evolving enhancer protein 3-2, chloroplastic\" & Source:UniProtKB/Swiss-Prot;Acc:Q41932]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mdf_name_uniprot\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'description'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'['\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mdf_name_uniprot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3022\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3023\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3024\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3025\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3026\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3080\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3082\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3084\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'description'"
     ]
    }
   ],
   "source": [
    "#Seperate descriptive name from uniprot symbol by first appearance of \"[\", expand in to two columns\n",
    "#e.g. \"Oxygen-evolving enhancer protein 3-2, chloroplastic [Source:UniProtKB/Swiss-Prot;Acc:Q41932]\" -> \"Oxygen-evolving enhancer protein 3-2, chloroplastic\" & Source:UniProtKB/Swiss-Prot;Acc:Q41932]\n",
    "\n",
    "df_name_uniprot = df_genes_aracore['description'].str.split('[', 1, expand=True,) \n",
    "df_name_uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9b3e8782",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_name_uniprot' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-b3b832ebccf2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#Merge with df_genes_aracore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf_genes_aracore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_name_uniprot\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m#rename columns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'gene_name'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'uniprot_id'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_name_uniprot' is not defined"
     ]
    }
   ],
   "source": [
    "#Merge with df_genes_aracore\n",
    "df_genes_aracore = pd.concat([df_genes_aracore, df_name_uniprot], axis=1) \n",
    "\n",
    "#rename columns\n",
    "df_genes_aracore.rename(columns={0: 'gene_name', 1: 'uniprot_id'}, inplace=True)\n",
    "df_genes_aracore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "763f7c75",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'uniprot_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'uniprot_id'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-287046fb7e76>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#Extract Uniprot id by splitting on last appearance of ':', expand columns, but only keep the last column for the uniprot id, the rest is trash\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#e.g. Source:UniProtKB/Swiss-Prot;Acc:Q41932] -> \"Source:UniProtKB/Swiss-Prot;Acc\" & \"Q41932]\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'uniprot_id'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'uniprot_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrsplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m':'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mexpand\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#Remove closing \"]\" from uniprot id\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3022\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3023\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3024\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3025\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3026\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3080\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3082\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3084\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'uniprot_id'"
     ]
    }
   ],
   "source": [
    "#Extract Uniprot id by splitting on last appearance of ':', expand columns, but only keep the last column for the uniprot id, the rest is trash\n",
    "#e.g. Source:UniProtKB/Swiss-Prot;Acc:Q41932] -> \"Source:UniProtKB/Swiss-Prot;Acc\" & \"Q41932]\"\n",
    "df_genes_aracore['uniprot_id'] = df_genes_aracore['uniprot_id'].str.rsplit(':',1,expand=True).iloc[:,-1]\n",
    "\n",
    "#Remove closing \"]\" from uniprot id \n",
    "#e.g. \"Q41932]\" -> \"Q41932\"\n",
    "df_genes_aracore['uniprot_id'] = df_genes_aracore['uniprot_id'].apply(lambda x: x[:-1] if isinstance(x,str) else None)\n",
    "df_genes_aracore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3c6bc2dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aracore_ids</th>\n",
       "      <th>aracore_name</th>\n",
       "      <th>aracore_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [aracore_ids, aracore_name, aracore_annotations]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_aracore[df_genes_aracore.isna().sum(axis=1) == 8] #2 Genes have no annotations at all -> Look-up those genes, maybe the Gene ids are wrong!?!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4657b4b6",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'ncbigene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ncbigene'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-cdce47eae23d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'ncbigene'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# ~28 genes without NCBI Symbol\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3022\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3023\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3024\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3025\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3026\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3080\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3082\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3084\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ncbigene'"
     ]
    }
   ],
   "source": [
    "df_genes_aracore[df_genes_aracore['ncbigene'].isna()] # ~28 genes without NCBI Symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e4ebd8a4",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'symbol'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'symbol'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-740a4a5d167b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'symbol'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# 97 genes have no official gene #symbol\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3022\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3023\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3024\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3025\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3026\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3080\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3082\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3084\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'symbol'"
     ]
    }
   ],
   "source": [
    "df_genes_aracore[df_genes_aracore['symbol'].isna()].head(100) # 97 genes have no official gene #symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8b553938",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'uniprot_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'uniprot_id'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-fa606cf230bd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'uniprot_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m#2 genes have no uniprot id (same as above)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3022\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3023\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3024\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3025\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3026\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3080\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3082\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3084\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'uniprot_id'"
     ]
    }
   ],
   "source": [
    "df_genes_aracore[df_genes_aracore['uniprot_id'].isna()] #2 genes have no uniprot id (same as above)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d211e0b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export final mapping table for manual mapping\n",
    "df_genes_aracore.to_csv('../data/processed/2021-05-27-genes-mapping-table.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
