{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b9a181a9",
   "metadata": {},
   "source": [
    "# Creating genes mapping table "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3aa01bdf",
   "metadata": {},
   "source": [
    "## I / Import the resources : AraCore model and ModelSEED and BiGG tables "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67020030",
   "metadata": {},
   "source": [
    "### 1) Install missing modules and load modules "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "feedbc17",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a370783c",
   "metadata": {},
   "source": [
    "### 2) Load AraCore Model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d512d2f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get file from github\n",
    "fileName = 'https://raw.githubusercontent.com/ma-blaetke/CBM_C3_C4_Metabolism/master/data/2018-23-05-mb-genC3.sbml'\n",
    "r = requests.get(fileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0acd412e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>c3_model</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fbe597fb4c0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>413</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>572</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Ex_Suc - 1.0*Ex_Suc_reverse_fb96e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Chloroplast, Lumen, Cytosol, Mitochondrion, IntermembraneSpace, Peroxisome</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model c3_model at 0x7fbe597fb4c0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create model\n",
    "model = cobra.io.read_sbml_model(r.text)\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d127becc",
   "metadata": {},
   "source": [
    "### 3) Correct gene naming in AraCore Model according to BiGG naming conventions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a220429",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2989798f",
   "metadata": {},
   "source": [
    "###  4) Correct gene naming in AraCore Model according to ModelSEED naming conventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c55790ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "de9877da",
   "metadata": {},
   "source": [
    "### 5) Create Gene Table for AraCore Model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3ee21b96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aracore_ids</th>\n",
       "      <th>aracore_name</th>\n",
       "      <th>aracore_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2G30570</td>\n",
       "      <td>AT2G30570</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATCG00560</td>\n",
       "      <td>ATCG00560</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT4G21280</td>\n",
       "      <td>AT4G21280</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATCG00300</td>\n",
       "      <td>ATCG00300</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>AT3G27620</td>\n",
       "      <td>AT3G27620</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>AT5G64210</td>\n",
       "      <td>AT5G64210</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>801</th>\n",
       "      <td>AT1G32350</td>\n",
       "      <td>AT1G32350</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>802</th>\n",
       "      <td>AT3G22360</td>\n",
       "      <td>AT3G22360</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>AT4G22260</td>\n",
       "      <td>AT4G22260</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>804 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    aracore_ids aracore_name aracore_annotations\n",
       "0     AT4G05180    AT4G05180                  {}\n",
       "1     AT2G30570    AT2G30570                  {}\n",
       "2     ATCG00560    ATCG00560                  {}\n",
       "3     AT4G21280    AT4G21280                  {}\n",
       "4     ATCG00300    ATCG00300                  {}\n",
       "..          ...          ...                 ...\n",
       "799   AT3G27620    AT3G27620                  {}\n",
       "800   AT5G64210    AT5G64210                  {}\n",
       "801   AT1G32350    AT1G32350                  {}\n",
       "802   AT3G22360    AT3G22360                  {}\n",
       "803   AT4G22260    AT4G22260                  {}\n",
       "\n",
       "[804 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_aracore = pd.DataFrame(\n",
    "    {\n",
    "        \"aracore_ids\":[x.id for x in model.genes],\n",
    "        \"aracore_name\":[x.name for x in model.genes],\n",
    "        \"aracore_annotations\":[x.annotation for x in model.genes]\n",
    "    })\n",
    "df_genes_aracore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5b0f81b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "804"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_aracore[\"aracore_ids\"].shape[0] # indicates the number of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5492e583",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 804 entries, 0 to 803\n",
      "Data columns (total 3 columns):\n",
      " #   Column               Non-Null Count  Dtype \n",
      "---  ------               --------------  ----- \n",
      " 0   aracore_ids          804 non-null    object\n",
      " 1   aracore_name         804 non-null    object\n",
      " 2   aracore_annotations  804 non-null    object\n",
      "dtypes: object(3)\n",
      "memory usage: 19.0+ KB\n"
     ]
    }
   ],
   "source": [
    "df_genes_aracore.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd1654a",
   "metadata": {},
   "source": [
    "When we look at the metabolites or the reactions in the BiGG database, for example, we can notice that there are the correspondong genes, as well as their ID in the NCBI database. \n",
    "=> Check in the NCBI database : \n",
    "We have : \n",
    "- the gene ID\n",
    "- summary\n",
    "- the genomic context\n",
    "- the genomic regions, transcripts, and products\n",
    "- bibliography\n",
    "- the general protein information\n",
    "- NCBI preferenced sequence\n",
    "- related sequence\n",
    "- additional links\n",
    "=> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9dd94e7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'end': 2673243,\n",
       " 'logic_name': 'araport11',\n",
       " 'start': 2671523,\n",
       " 'db_type': 'core',\n",
       " 'assembly_name': 'TAIR10',\n",
       " 'id': 'AT4G05180',\n",
       " 'seq_region_name': '4',\n",
       " 'biotype': 'protein_coding',\n",
       " 'Transcript': [{'biotype': 'protein_coding',\n",
       "   'Exon': [{'id': 'AT4G05180.1.exon1',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'db_type': 'core',\n",
       "     'seq_region_name': '4',\n",
       "     'end': 2673243,\n",
       "     'start': 2672873,\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'object_type': 'Exon',\n",
       "     'strand': -1},\n",
       "    {'strand': -1,\n",
       "     'object_type': 'Exon',\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'start': 2672340,\n",
       "     'end': 2672635,\n",
       "     'seq_region_name': '4',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'id': 'AT4G05180.1.exon2',\n",
       "     'db_type': 'core'},\n",
       "    {'object_type': 'Exon',\n",
       "     'strand': -1,\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'end': 2672191,\n",
       "     'start': 2671900,\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'id': 'AT4G05180.1.exon3',\n",
       "     'db_type': 'core',\n",
       "     'seq_region_name': '4'},\n",
       "    {'strand': -1,\n",
       "     'object_type': 'Exon',\n",
       "     'species': 'arabidopsis_thaliana',\n",
       "     'start': 2671523,\n",
       "     'end': 2671718,\n",
       "     'seq_region_name': '4',\n",
       "     'db_type': 'core',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'id': 'AT4G05180.1.exon4'}],\n",
       "   'id': 'AT4G05180.1',\n",
       "   'db_type': 'core',\n",
       "   'start': 2671523,\n",
       "   'logic_name': 'araport11',\n",
       "   'strand': -1,\n",
       "   'is_canonical': 1,\n",
       "   'source': 'araport11',\n",
       "   'Parent': 'AT4G05180',\n",
       "   'Translation': {'species': 'arabidopsis_thaliana',\n",
       "    'length': 230,\n",
       "    'Parent': 'AT4G05180.1',\n",
       "    'object_type': 'Translation',\n",
       "    'id': 'AT4G05180.1',\n",
       "    'db_type': 'core',\n",
       "    'start': 2672093,\n",
       "    'end': 2673170},\n",
       "   'seq_region_name': '4',\n",
       "   'assembly_name': 'TAIR10',\n",
       "   'end': 2673243,\n",
       "   'species': 'arabidopsis_thaliana',\n",
       "   'object_type': 'Transcript'},\n",
       "  {'species': 'arabidopsis_thaliana',\n",
       "   'object_type': 'Transcript',\n",
       "   'seq_region_name': '4',\n",
       "   'assembly_name': 'TAIR10',\n",
       "   'end': 2673243,\n",
       "   'is_canonical': 0,\n",
       "   'source': 'araport11',\n",
       "   'strand': -1,\n",
       "   'Translation': {'end': 2673170,\n",
       "    'start': 2672093,\n",
       "    'id': 'AT4G05180.2',\n",
       "    'db_type': 'core',\n",
       "    'Parent': 'AT4G05180.2',\n",
       "    'object_type': 'Translation',\n",
       "    'length': 230,\n",
       "    'species': 'arabidopsis_thaliana'},\n",
       "   'Parent': 'AT4G05180',\n",
       "   'Exon': [{'end': 2673243,\n",
       "     'start': 2672873,\n",
       "     'id': 'AT4G05180.1.exon1',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'db_type': 'core',\n",
       "     'seq_region_name': '4',\n",
       "     'object_type': 'Exon',\n",
       "     'strand': -1,\n",
       "     'species': 'arabidopsis_thaliana'},\n",
       "    {'end': 2672635,\n",
       "     'start': 2672340,\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'id': 'AT4G05180.1.exon2',\n",
       "     'db_type': 'core',\n",
       "     'seq_region_name': '4',\n",
       "     'object_type': 'Exon',\n",
       "     'strand': -1,\n",
       "     'species': 'arabidopsis_thaliana'},\n",
       "    {'end': 2672191,\n",
       "     'start': 2671784,\n",
       "     'db_type': 'core',\n",
       "     'id': 'AT4G05180.2.exon3',\n",
       "     'assembly_name': 'TAIR10',\n",
       "     'seq_region_name': '4',\n",
       "     'object_type': 'Exon',\n",
       "     'strand': -1,\n",
       "     'species': 'arabidopsis_thaliana'}],\n",
       "   'biotype': 'protein_coding',\n",
       "   'db_type': 'core',\n",
       "   'id': 'AT4G05180.2',\n",
       "   'start': 2671784,\n",
       "   'logic_name': 'araport11'}],\n",
       " 'object_type': 'Gene',\n",
       " 'display_name': 'PSBQ2',\n",
       " 'source': 'araport11',\n",
       " 'strand': -1,\n",
       " 'species': 'arabidopsis_thaliana',\n",
       " 'description': 'Oxygen-evolving enhancer protein 3-2, chloroplastic [Source:UniProtKB/Swiss-Prot;Acc:Q41932]'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#EXAMPLE REQUEST:\n",
    "\n",
    "#Retrieve Information on gene symbol, location (chromosome, start and end), uniprot id/descriptive name using the ENSEMBL REST API\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/lookup\n",
    "\n",
    "gene_id = 'AT4G05180'\n",
    "\n",
    "#ENSEMBL Server Url\n",
    "server = \"https://rest.ensembl.org\"\n",
    "\n",
    "#URL specification to get meta-information of a gene by its id\n",
    "ext = f\"/lookup/id/{gene_id}?expand=1\"\n",
    "  \n",
    "#Request information from URL in json format\n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "\n",
    "#Check if request was successful\n",
    "if r.ok:\n",
    "    \n",
    "    #Get json (dictionary-like structure) from request\n",
    "    json_anno = r.json()\n",
    "\n",
    "else:\n",
    "\n",
    "    json_anno = None\n",
    "\n",
    "json_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2f032d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retrieve Information on gene symbol, location (chromosome, start and end), uniprot id/descriptive name using the ENSEMBL REST API\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/lookup\n",
    "\n",
    "def lookup_ensembl_id(gene_id):\n",
    "\n",
    "  #pandas series to hold information\n",
    "  ds_anno = pd.Series(index=['symbol','chr','start', 'end', 'description'])\n",
    "  \n",
    "  #ENSEMBL Server Url\n",
    "  server = \"https://rest.ensembl.org\"\n",
    "  \n",
    "  #URL specification to get meta-information of a gene by its id\n",
    "  ext = f\"/lookup/id/{gene_id}?expand=1\"\n",
    "  \n",
    "  #Request information from URL in json format\n",
    "  r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "  \n",
    "  #Check if request wad successful\n",
    "if r.ok:\n",
    "    \n",
    "    #Get json (dictionary-like structure) from request\n",
    "    json_anno = r.json()\n",
    "\n",
    "    #Save relevant information in pandas series\n",
    "    ds_anno['symbol'] = json_anno.get('display_name', None)\n",
    "    ds_anno['chr'] = json_anno.get('seq_region_name', None)\n",
    "    ds_anno['start'] = json_anno.get('start', None)\n",
    "    ds_anno['end'] = json_anno.get('end', None)\n",
    "    ds_anno['description'] = json_anno.get('description', None)\n",
    "else:\n",
    "    #Save set pandas series to None if request was not successfull\n",
    "    ds_anno['symbol'] = None\n",
    "    ds_anno['chr'] = None\n",
    "    ds_anno['start'] = None\n",
    "    ds_anno['end'] = None\n",
    "    ds_anno['description'] = None\n",
    "\n",
    "  #return pandas series\n",
    "  return ds_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9e75d806",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db_display_name</th>\n",
       "      <th>primary_id</th>\n",
       "      <th>description</th>\n",
       "      <th>display_id</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>version</th>\n",
       "      <th>info_text</th>\n",
       "      <th>info_type</th>\n",
       "      <th>ensembl_end</th>\n",
       "      <th>xref_start</th>\n",
       "      <th>ensembl_start</th>\n",
       "      <th>ensembl_identity</th>\n",
       "      <th>score</th>\n",
       "      <th>xref_identity</th>\n",
       "      <th>cigar_line</th>\n",
       "      <th>evalue</th>\n",
       "      <th>xref_end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NASC_GENE_ID</th>\n",
       "      <td>NASC Gene ID</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>photosystem II subunit Q-2</td>\n",
       "      <td>AT4G05180-TAIR-G</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ArrayExpress</th>\n",
       "      <td>Expression Atlas</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td></td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAIR_LOCUS</th>\n",
       "      <td>TAIR</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>photosystem II subunit Q-2</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAIR_SYMBOL</th>\n",
       "      <td>TAIR Gene Name</td>\n",
       "      <td>PSBQ</td>\n",
       "      <td></td>\n",
       "      <td>PSBQ</td>\n",
       "      <td>[PSBQ-1, PSBQ-2, PSBQA, PSII-Q]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniGene</th>\n",
       "      <td>UniGene</td>\n",
       "      <td>At.21853</td>\n",
       "      <td>Oxygen-evolving enhancer protein 3-2</td>\n",
       "      <td>At.21853</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>SEQUENCE_MATCH</td>\n",
       "      <td>1037.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>5185.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1037M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1037.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EntrezGene</th>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>825866</td>\n",
       "      <td>None</td>\n",
       "      <td>825866</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>DEPENDENT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KNETMINER_ARA</th>\n",
       "      <td>KNETMINER_ARA</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>None</td>\n",
       "      <td>AT4G05180</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>DEPENDENT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               db_display_name primary_id  \\\n",
       "dbname                                                      \n",
       "NASC_GENE_ID                      NASC Gene ID  AT4G05180   \n",
       "ArrayExpress                  Expression Atlas  AT4G05180   \n",
       "TAIR_LOCUS                                TAIR  AT4G05180   \n",
       "TAIR_SYMBOL                     TAIR Gene Name       PSBQ   \n",
       "UniGene                                UniGene   At.21853   \n",
       "EntrezGene     NCBI gene (formerly Entrezgene)     825866   \n",
       "KNETMINER_ARA                    KNETMINER_ARA  AT4G05180   \n",
       "\n",
       "                                        description        display_id  \\\n",
       "dbname                                                                  \n",
       "NASC_GENE_ID             photosystem II subunit Q-2  AT4G05180-TAIR-G   \n",
       "ArrayExpress                                                AT4G05180   \n",
       "TAIR_LOCUS               photosystem II subunit Q-2         AT4G05180   \n",
       "TAIR_SYMBOL                                                      PSBQ   \n",
       "UniGene        Oxygen-evolving enhancer protein 3-2          At.21853   \n",
       "EntrezGene                                     None            825866   \n",
       "KNETMINER_ARA                                  None         AT4G05180   \n",
       "\n",
       "                                      synonyms version info_text  \\\n",
       "dbname                                                             \n",
       "NASC_GENE_ID                                []       0             \n",
       "ArrayExpress                                []       0             \n",
       "TAIR_LOCUS                                  []       0             \n",
       "TAIR_SYMBOL    [PSBQ-1, PSBQ-2, PSBQA, PSII-Q]       0             \n",
       "UniGene                                     []       0             \n",
       "EntrezGene                                  []       0             \n",
       "KNETMINER_ARA                               []       0             \n",
       "\n",
       "                    info_type  ensembl_end  xref_start  ensembl_start  \\\n",
       "dbname                                                                  \n",
       "NASC_GENE_ID           DIRECT          NaN         NaN            NaN   \n",
       "ArrayExpress           DIRECT          NaN         NaN            NaN   \n",
       "TAIR_LOCUS             DIRECT          NaN         NaN            NaN   \n",
       "TAIR_SYMBOL            DIRECT          NaN         NaN            NaN   \n",
       "UniGene        SEQUENCE_MATCH       1037.0         1.0            1.0   \n",
       "EntrezGene          DEPENDENT          NaN         NaN            NaN   \n",
       "KNETMINER_ARA       DEPENDENT          NaN         NaN            NaN   \n",
       "\n",
       "               ensembl_identity   score  xref_identity cigar_line  evalue  \\\n",
       "dbname                                                                      \n",
       "NASC_GENE_ID                NaN     NaN            NaN        NaN     NaN   \n",
       "ArrayExpress                NaN     NaN            NaN        NaN     NaN   \n",
       "TAIR_LOCUS                  NaN     NaN            NaN        NaN     NaN   \n",
       "TAIR_SYMBOL                 NaN     NaN            NaN        NaN     NaN   \n",
       "UniGene                    96.0  5185.0          100.0      1037M     NaN   \n",
       "EntrezGene                  NaN     NaN            NaN        NaN     NaN   \n",
       "KNETMINER_ARA               NaN     NaN            NaN        NaN     NaN   \n",
       "\n",
       "               xref_end  \n",
       "dbname                   \n",
       "NASC_GENE_ID        NaN  \n",
       "ArrayExpress        NaN  \n",
       "TAIR_LOCUS          NaN  \n",
       "TAIR_SYMBOL         NaN  \n",
       "UniGene          1037.0  \n",
       "EntrezGene          NaN  \n",
       "KNETMINER_ARA       NaN  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#EXAMPLE REQUEST\n",
    "\n",
    "#Retrieve Information on cross references for each gene symbol (NCBI gene id)\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/xref_id\n",
    "\n",
    "gene_id = 'AT4G05180'\n",
    "\n",
    "#ENSEMBL Server Url\n",
    "server = \"https://rest.ensembl.org\"\n",
    "\n",
    "#URL specification to get meta-information of a gene by its id\n",
    "ext = f\"/xrefs/id/{gene_id}?\"\n",
    "  \n",
    "#Request information from URL in json format\n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "\n",
    "#Check if request wad successful\n",
    "if r.ok:\n",
    "    \n",
    "    #Get json (dictionary-like structure) from request\n",
    "    json_anno = r.json()\n",
    "\n",
    "    #Convert json to pandas dataframe format to easily retrieve information\n",
    "    df_anno = pd.DataFrame(json_anno)\n",
    "    df_anno.set_index('dbname', inplace=True)\n",
    "  \n",
    "else:\n",
    "     df_anno = None\n",
    "\n",
    "df_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "039ea24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retrieve Information on cross references for each gene symbol (NCBI gene id)\n",
    "#https://rest.ensembl.org/\n",
    "#https://rest.ensembl.org/documentation/info/xref_id\n",
    "\n",
    "def xrefs_ensembl_id(gene_id):\n",
    "  \n",
    "  #pandas series to hold information\n",
    "  ds_anno = pd.Series(index=['ncbigene'])\n",
    "  \n",
    "  #ENSEMBL Server Url\n",
    "  server = \"https://rest.ensembl.org\"\n",
    "  #URL specification to get meta-information of a gene by its id\n",
    "  ext = f\"/xrefs/id/{gene_id}?\"\n",
    "  \n",
    "  #Request information from URL in json format\n",
    "  r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "  \n",
    "  #Check if request wad successful\n",
    "  if r.ok:\n",
    "    \n",
    "    #Get json (dictionary-like structure) from request\n",
    "    json_anno = r.json()\n",
    "\n",
    "    #Convert json to pandas dataframe format\n",
    "    df_anno = pd.DataFrame(json_anno)\n",
    "    df_anno.set_index('dbname', inplace=True)\n",
    "    \n",
    "    #Save relevant information in pandas series\n",
    "    if 'EntrezGene' in df_anno.index:\n",
    "      if 'primary_id' in df_anno.columns:\n",
    "        ds_anno['ncbigene'] = str(df_anno.loc['EntrezGene','primary_id'])\n",
    "  \n",
    "  else:\n",
    "     ds_anno['ncbigene'] = None\n",
    " \n",
    "  return ds_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8e030cf9",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Series' object has no attribute 'parallel_apply'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-a3f59e75c06b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#Request information for each gene using xrefs_ensembl_id() function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf_genes_xref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_genes_aracore\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'aracore_ids'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparallel_apply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mxrefs_ensembl_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/photosynthesis_bis/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5463\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5464\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5465\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5466\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5467\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Series' object has no attribute 'parallel_apply'"
     ]
    }
   ],
   "source": [
    "#Request information for each gene using xrefs_ensembl_id() function\n",
    "df_genes_xref = df_genes_aracore['aracore_ids'].parallel_apply(lambda x: xrefs_ensembl_id(x)) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d63ddafd",
   "metadata": {},
   "source": [
    "### 6) Import BiGG genes file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3faf99f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "92607592",
   "metadata": {},
   "source": [
    "### 7) Import ModelSEED genes file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c336ae4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "11af7b14",
   "metadata": {},
   "source": [
    "## II / Extract the annotations from the AraCore model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6adb025a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a96e6a67",
   "metadata": {},
   "source": [
    "## III / Extract the id from the annotations we extracted previously (2nd step) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b057f875",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d6aa5d61",
   "metadata": {},
   "source": [
    "## IV / Merge AraCore table with BiGG one "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "139334e4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
